pkg/recovery/apiserver.go

```diff
@@ -0,0 +1,438 @@
+					ClientKeyData:         restConfig.KeyData,
+				},
+			},
+		},
+	}, nil
+}
+
+func (s *Apiserver) GetKubeClientset() (*kubernetes.Clientset, error) {
+	if s.kubeClientSet != nil {
+		return s.kubeClientSet, nil
+	}
+
+	restConfig, err := s.RestConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	kubeClientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build kubernetes clientset: %v", err)
+	}
+
+	s.kubeClientSet = kubeClientset
+
+	return s.kubeClientSet, nil
+}
+
+func (s *Apiserver) recoveryPod() (*corev1.Pod, error) {
+	recoveryDirPath, err := s.GetRecoveryDirPath()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get recovery dir: %v", err)
+	}
+
+	// Create the manifest to run recovery apiserver
+	recoveryPodTemplateBytes, err := v311_00_assets.Asset(RecoveryPodAsset)
+	if err != nil {
+		return nil, fmt.Errorf("fail to find internal recovery pod asset %q: %v", RecoveryPodAsset, err)
+	}
+
+	// Process the template
+	t, err := template.New("recovery-pod-template").Parse(string(recoveryPodTemplateBytes))
+	if err != nil {
+		return nil, fmt.Errorf("fail to parse internal recovery pod template %q: %v", RecoveryPodAsset, err)
+	}
+
+	recoveryPodBuffer := bytes.NewBuffer(nil)
+	err = t.Execute(recoveryPodBuffer, struct {
+		KubeApiserverImage string
+		ResourceDir        string
+	}{
+		KubeApiserverImage: s.KubeApiserverImage,
+		ResourceDir:        recoveryDirPath,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("fail to execute internal recovery pod template %q: %v", RecoveryPodAsset, err)
+	}
+
+	recoveryPodObj, err := runtime.Decode(Codecs.UniversalDecoder(corev1.SchemeGroupVersion), recoveryPodBuffer.Bytes())
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode internal recovery pod %q: %v", RecoveryPodAsset, err)
+	}
+
+	recoveryPod, ok := recoveryPodObj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("unsupported type: internal recovery pod is not type *corev1.Pod but %T", recoveryPod)
+	}
+
+	return recoveryPod, nil
+}
+
+func (s *Apiserver) Create() error {
+	resourceDirPath, err := s.GetResourceDirPath()
+	if err != nil {
+		return fmt.Errorf("failed to get resource dir: %v", err)
+	}
+
+	recoveryDirPath, err := s.GetRecoveryDirPath()
+	if err != nil {
+		return fmt.Errorf("failed to get recovery dir: %v", err)
+	}
+
+	// Copy certs for accessing etcd
+	for src, dest := range map[string]string{
+		"secrets/etcd-client/tls.key":              "etcd-client.key",
+		"secrets/etcd-client/tls.crt":              "etcd-client.crt",
+		"configmaps/etcd-serving-ca/ca-bundle.crt": "etcd-serving-ca-bundle.crt",
+	} {
+		err = copyFile(path.Join(resourceDirPath, src), path.Join(recoveryDirPath, dest))
+		if err != nil {
+			return err
+		}
+	}
+
+	// We are creating only temporary certificates to start the recovery apiserver.
+	// A week seem reasonably high for a debug session, while it is easy to create a new one.
+	certValidity := 7 * 24 * time.Hour
```

Comment:
this would deserve some log output maybe: "Creating temporary certs with 1 week validity"